sled-agent/src/hardware/illumos/mod.rs

```diff
@@ -73,56 +170,188 @@ fn node_name(subsystem_vid: i32, subsystem_id: i32) -> String {
+            return Err(Error::NoDevLinks(PathBuf::from(devfs_path)));
+        }
+        return Ok(Some(paths[0].path().to_path_buf()));
+    }
+    Ok(None)
+}
+
+// Gather a notion of "device identity".
+fn blkdev_device_identity(devfs_path: &str) -> Result<DiskIdentity, Error> {
+    let path = PathBuf::from(devfs_path);
+    let ctrl_id = nvme::identify_controller(&path)
+        .map_err(|err| Error::MissingNVMEIdentity { path, err })?;
+
+    let vendor =
+        format!("{:x}{:x}", ctrl_id.vendor_id, ctrl_id.subsystem_vendor_id);
+    let serial = ctrl_id.serial;
+    let model = ctrl_id.model;
+    Ok(DiskIdentity { vendor, serial, model })
+}
+
+fn poll_blkdev_node(
+    log: &Logger,
+    disks: &mut HashSet<UnparsedDisk>,
+    mut node: Node<'_>,
+) -> Result<(), Error> {
+    let Some(driver_name) = node.driver_name() else {
+        return Ok(());
+    };
+
+    if driver_name != "blkdev" {
+        return Ok(());
+    }
+
+    let devfs_path = node.devfs_path().map_err(Error::DevInfo)?;
+    let dev_path = get_dev_path_of_whole_disk(&node)?;
+
+    // For some reason, libdevfs doesn't prepend "/devices" when
+    // referring to the path, but it still returns an absolute path.
+    //
+    // Validate that we're still using this leading slash, but make the
+    // path "real".
+    assert!(devfs_path.starts_with('/'));
+    let devfs_path = format!("/devices{devfs_path}");
+    let mut device_id = None;
+    while let Some(parent) = node.parent().map_err(Error::DevInfo)? {
+        node = parent;
+
+        if let Some(driver_name) = node.driver_name() {
+            if driver_name == "nvme" {
+                device_id = Some(blkdev_device_identity(&format!(
+                    "/devices{}:devctl",
+                    node.devfs_path().map_err(Error::DevInfo)?
+                ))?);
+            }
+        }
+        let slot_prop_name = "physical-slot#".to_string();
+        if let Some(Ok(slot_prop)) = node.props().into_iter().find(|prop| {
+            if let Ok(prop) = prop {
+                prop.name() == slot_prop_name
+            } else {
+                false
+            }
+        }) {
+            let slot = slot_prop.as_i64().ok_or_else(|| {
+                Error::UnexpectedPropertyType {
+                    name: slot_prop_name,
+                    ty: "i64".to_string(),
+                }
+            })?;
+
+            let Some(variant) = slot_to_disk_variant(slot) else {
+                warn!(log, "Slot# {slot} is not recognized as a disk: {devfs_path}");
+                break;
+            };
+            let disk = UnparsedDisk::new(
+                PathBuf::from(&devfs_path),
+                dev_path,
+                slot,
+                variant,
+                device_id.ok_or_else(|| {
+                    Error::NotAnNVMEDevice(PathBuf::from(devfs_path))
+                })?,
+            );
+            disks.insert(disk);
+            break;
+        }
+    }
```

Comment:
I think rather than a loop, here, which seems like it would admit some tree structures that are not quite what we are expecting, I would do an explicit walk up the tree looking for the precise shape we want; e.g., the sketch below:

```rust
/*
 * Locate nvme parent of blkdev node:
 */
let Some(p) = node.parent()? else {
    bail!("blkdev has no parent device?");
};
if p.driver_name.as_deref() != Some("nvme") {
    bail!("expected nvme parent, got {:?}", p.driver_name);
}
/*
 * XXX Get the inquiry-* properties from the node while we're here.
 */

/*
 * Locate the PCI parent of the nvme node:
 */
let Some(p) = p.parent()? else {
    bail!("nvme has no parent device?");
};
let slot = p.props()
    .filter_map(|res| res.ok())
    .filter(|prop| prop.name() == "physical-slot#")
    .filter_map(|prop| {
        let num: Option<u32> = prop.as_i64().try_into();
        num
    })
    .next();
let Some(slot) = slot else {
    bail!("no physical slot number on PCI bridge?");
};
let Some(v) = slot_to_disk_variant(slot) else {
    bail!("unrecognised disk slot {}", slot);
};

let disk = UnparsedDisk::new(
...
```

Reply:
Looks great!